### PR TITLE
Implement customizable RoPE

### PR DESCRIPTION
**Pull request by jxy**
The original RoPE has pre-defined parameters

theta_i = 10000^(−2(i−1)/d), for i in [1, 2, ..., d/2]

Our customizable RoPE, ggml_rope_custom_inplace, uses

theta_i = scale * base^(−2(i−1)/d), for i in [1, 2, ..., d/2]

with the default matches the original

scale = 1.0
base = 10000

The new command line arguments
--rope-freq-base
--rope-freq-scale
set the two new RoPE parameter.

Recent researches show changing these two parameters extends the context limit with minimal loss.

1. Extending Context to 8K kaiokendev https://kaiokendev.github.io/til#extending-context-to-8k

2. Extending Context Window of Large Language Models via Positional Interpolation Shouyuan Chen, Sherman Wong, Liangjian Chen, Yuandong Tian https://arxiv.org/abs/2306.15595

3. NTK-Aware Scaled RoPE allows LLaMA models to have extended (8k+) context size without any fine-tuning and minimal perplexity degradation. https://www.reddit.com/user/bloc97 https://www.reddit.com/r/LocalLLaMA/comments/14lz7j5/ntkaware_scaled_rope_allows_llama_models_to_have/

For the bold, try adding the following command line parameters to your favorite model: -c 16384 --rope-freq-base 80000 --rope-freq-scale 0.5

### CODE REVIEW

**Comment by ymcui** [comment]
Hello, does this PR support `./perplexity` for benchmarking? 
I did not find significant PPL decrease with this PR compared to the main branch code with 6K/8K ctx.
Command used:
```
./perplexity -m zh-llama-models/p7b/ggml-model-q6_k.bin -f text.txt -ngl 1 --rope-freq-scale 1.0 --rope-freq-base 10000 -c 6144
```

**Comment by 1980Dragon** [comment]
Just came across another RoPE adjustment method on Reddit. Thought it might be helpful, so here's the link!
[Dynamically Scaled RoPE further increases performance of long context LLaMA with zero fine-tuning](https://www.reddit.com/r/LocalLLaMA/comments/14mrgpr/dynamically_scaled_rope_further_increases/)

image

**Comment by maddes8cht** [comment]
This still means i will get better perplexity-performance when usingvthis pr with @TheBloke 's supehot model-variants I guess?

**Comment by FNsi** [comment]
Somehow exciting? 

I use my merged 13b model (wizard vicuña + starcoder + superhot 16k) 

With that 16k command.

1. 5.3710, 2. 5.7513 

Looks reasonable.

**Comment by jxy** [comment]
The `--rope-freq-scale` is the same scale used in "superhot/rope interpolation".  superhot 8k lora corresponds to `--rope-freq-scale 0.25 -c 8192`, which is a factor of 4 increase. Similarly superhot 16k lora and longchat 16k corresponds to `--rope-freq-scale 0.125 -c 16384`, for a factor of 8.

The `--rope-freq-base` simplifies the "NTK-Aware Scaled RoPE". The base number here corresponds to `10000*alpha**(64/63)` using the `alpha` introduced in the reddit post.  I'm not aware of any direct translation of how context length corresponds to the `base` or `alpha`.  My limited testing with 13B models show a rough quadratic correspondence, `C = -0.0652*b*b + 0.862*b + 0.203`, for `C` the factor of context length increase, and `b` the factor of base increase, roughly

base | effective ctx factor | effective ctx
-|-|-
20000 | 1.66 | 3400
26000 | 2 | 4096
40000 | 2.6 | 5300
57200 | 3.0 | 6144

I found `base>60000` didn't feel good, though I've no hard numbers to back this up.

Empirically, without fine tune, you could try
- `-c 4096 --rope-freq-scale 0.83 --rope-freq-base 20000`
- `-c 6144 --rope-freq-scale 0.86 --rope-freq-base 40000`
- `-c 8192 --rope-freq-scale 0.75 --rope-freq-base 57200`

With superhot 16k or longchat, perhaps you could try (KV cache along requires 25GB!!)
- `-c 32768 --rope-freq-scale 0.125 --rope-freq-base 26000`
- or dial up base more?

**Comment by jxy** [comment]
I used some numbers posted by @JohannesGaessler, and made changes in scratch0 size in this PR. I can rebase this PR on their PR #2056 if needed.

**Comment by JohannesGaessler** [comment]
I think the numbers that I determined for the VRAM scratch buffer will probably work for the RAM scratch buffer but I would still advise you to be cautious since the two types of scratch buffer work differently (the VRAM scratch buffer has tighter limits).

**Comment by trap20** [comment]
I tried the `-c 4096 --rope-freq-scale 0.83 --rope-freq-base 20000` configuration with the `wizardlm-33b-v1.0-uncensored.ggmlv3.q5_K_M.bin` model and got `ggml_new_tensor_impl: not enough space in the scratch memory pool (needed 490733568, available 482344960)`. Even with 0 layers offloaded.

**Comment by jxy** [comment]
> I tried the `-c 4096 --rope-freq-scale 0.83 --rope-freq-base 20000` configuration with the `wizardlm-33b-v1.0-uncensored.ggmlv3.q5_K_M.bin` model and got `ggml_new_tensor_impl: not enough space in the scratch memory pool (needed 490733568, available 482344960)`.

I can't reproduce. Is CUDA build different? I can't test CUDA build. The number 482344960 seems to be computed from 
```
{ MODEL_30B, ((size_t) n_ctx / 10ull + 256ull) * MB }
```
with `n_ctx = 2048`.  Do you use `main` or other method?

**Comment by time-less-ness** [comment]
I am trying this out, and it is working fine for me so far, though I've only tried:

```
gpt4-alpaca-lora-30b.ggmlv3.q4_0.bin
-c 6144 --rope-freq-scale 0.86 --rope-freq-base 40000
```

I'll keep trying other models and context sizes and seeing how it goes. Not sure if other fixes are included, but this seems to make inference way faster (via fewer long pauses to do CPU-related tasks) on my machine as well. Possibly just due to not having to recompute context as I hit the 2048-byte mark so often?

**Comment by trap20** [comment]
> I can't reproduce. Is CUDA build different?[...} Do you use `main` or other method/executable to run it?

Yes, it's a CUDA build for a 1080ti and I really should have used `main` for reporting, but didn't. I'll try `main` tomorrow.

**Comment by jxy** [comment]
> > I can't reproduce. Is CUDA build different?[...} Do you use `main` or other method/executable to run it?
> 
> Yes, it's a CUDA build for a 1080ti and I really should have used `main` for reporting, but didn't. I'll try `main` tomorrow.

If you could give me a stack trace of when MEM_REQ_SCRATCH0 is called, I could try to figure out what is wrong with the CUDA build. Otherwise, I'll see if I can get a system somewhere with cuda.

**Comment by trap20** [comment]
Can't reproduce the error today, no idea what I did exactly to trigger it...

**Comment by ggerganov** [comment]
This looks great, but similar to #1967 - let's wait for a while before merging.
There are new RoPE scaling techniques popping up by the hour each one better the other. No reason to commit to something just yet

**Comment by FNsi** [comment]
> - or dial up base more?

No good result by -c 16384 --rope-freq-scale 0.5 --rope-freq-base 57200

With perplexity 
[1] 6.7699

But by base 80000,

[1] 5.3710

So yes, dial up base is a good way.

**Comment by FNsi** [comment]
Running perplexity Openllama 3b  with -C 16384 scale 0.5 base 90000

No enough space in the contexts memory pool needed 543758112 available 536870912

**Comment by SlyEcho** [comment]
> KV cache alone requires 25GB!!

Could we quantize the KV cache?

**Comment by FNsi** [comment]
> > KV cache alone requires 25GB!!
> 
> 
> 
> Could we quantize the KV cache?

Try? ？ but how to start it?

**Comment by Green-Sky** [comment]
> > KV cache alone requires 25GB!!
> 
> Could we quantize the KV cache?

I think this was tried, and resulted in bad results. It should already be in f16.
but i dont remember, if we tried 8bit quantization...

**Comment by ardfork** [comment]
For some reason, the server example output some random unicode characters when using `--rope-freq-scale 0.25 -c 8192` but `--rope-freq-scale 0.25 -c 4096` work correctly and `--rope-freq-scale 0.25 -c 8192` work on cli.

**Comment by jxy** [comment]
server gives me 413 when the json data is large. We need help from those who contributed server code.

**Comment by SlyEcho** [comment]
> server gives me 413 when the json data is large. We need help from those who contributed server code.

I believe `CPPHTTPLIB_RECV_BUFSIZ` needs to be increased, right now it is 4K.

**Comment by digiwombat** [comment]
Yeah SlyEcho is right based on what I saw in the lib, setting
`#define CPPHTTPLIB_RECV_BUFSIZ size_t(<SOME NUMBER HERE>)`
before the httplib.h import should be the correct way to increase it, I believe.

**Comment by jxy** [comment]
> Running perplexity Openllama 3b with -C 16384 scale 0.5 base 90000
> 
> No enough space in the contexts memory pool needed 543758112 available 536870912

The only thing that I know of allocating 512 MB (536870912) is from `MEM_REQ_EVAL`, which this PR didn't change. Maybe try changing the line

```
{ MODEL_3B, 512ull * MB },
```

to something like

```
{ MODEL_3B, 600ull * MB },
```

and see if it helps?

**Comment by jxy** [comment]
> I believe `CPPHTTPLIB_RECV_BUFSIZ` needs to be increased, right now it is 4K.

It looks like a simple read buffer to me, and it's separate from the overall size limit.

**Comment by digiwombat** [comment]
> It looks like a simple read buffer to me, and it's separate from the overall size limit.

`Server::set_payload_max_length(uint64_t length)` might be what we're after then.

`svr.set_payload_max_length(1024 * 1024 * 1);` would set it to 1MB (left the 1 in for example purposes)

**Comment by FNsi** [comment]
> and see if it helps?

i think there's same problem in 13b while using 32k, 1024 limited. 

Btw. If we using base 160000 scale 0.25, I guess we can achieve 64k in reasonable response.

**Comment by jxy** [comment]
> > It looks like a simple read buffer to me, and it's separate from the overall size limit.
> 
> `Server::set_payload_max_length(uint64_t length)` might be what we're after then.
> 
> `svr.set_payload_max_length(1024 * 1024 * 1);` would set it to 1MB (left the 1 in for example purposes)

the default is actually
```
#define CPPHTTPLIB_PAYLOAD_MAX_LENGTH ((std::numeric_limits<size_t>::max)())
```

**Comment by jxy** [comment]
The issue with server is just because the curl commands don't have the proper content type specified, and that triggers a limit in httplib.h, `CPPHTTPLIB_FORM_URL_ENCODED_PAYLOAD_MAX_LENGTH 8192`.

**Comment by tmm1** [comment]
LLongMA https://twitter.com/enricoshippole/status/1677346583640256513?s=46&t=hIokEbug9Pr72tQFuXVULA

**Comment by Cebtenzzre** [comment]
[This comment](https://github.com/huggingface/transformers/pull/24653#issuecomment-1633027893) from the transformers PR helped me understand what this new LLongMAv2/NTK-By-Parts/NTKv2 method is about (they're all the same thing).

<details>
<summary>Graph from the scaled-rope PR comparing NTK to NTKv2 ("corrected")</summary>

graph 1

</details>

<details>
<summary>Graph from the transformers PR comparing Dynamic NTK to LLongMA and NTKv2</summary>

graph 2

</details>

<details>
<summary>Graph from an unknown source comparing several methods</summary>

graph 3

</details>

**Comment by SlyEcho** [comment]
I converted and quantized LLongMA-3b to ggml: [huggingface.co/SlyEcho/LLongMA-3b-ggml](https://huggingface.co/SlyEcho/LLongMA-3b-ggml).

It does seem to be working with long contexts using `-c 8192 --rope-freq-scale 0.25`.

**Comment by jxy** [comment]
> It'd probably be a good idea to investigate the changes made in [in this PR](https://github.com/jquesnelle/scaled-rope/pull/1). See also the author's comment on the [transformers PR](https://github.com/huggingface/transformers/pull/24653#issuecomment-1627627083)

> [This comment](https://github.com/huggingface/transformers/pull/24653#issuecomment-1633027893) from the transformers PR helped me understand what this new LLongMAv2/NTK-By-Parts/NTKv2 method is about (they're all the same thing).

Somebody please explain the python code in equations.

**Comment by JackJollimore** [comment]
> I converted and quantized LLongMA-3b to ggml: [huggingface.co/SlyEcho/LLongMA-3b-ggml](https://huggingface.co/SlyEcho/LLongMA-3b-ggml).
> 
> It does seem to be working with long contexts using `-c 8192 --rope-freq-scale 0.25`.

I _tried_ loading a .txt with 7200 tokens with 8192 context. It was going well, slowing down around 6k~ tokens, then somewhere near 7k~ my O.S. crashed(Android).

It seems that my device ran of Ram before completing the load, so I'll try again later with half as many tokens :)

**Comment by ggerganov** [comment]
What is the latest state of this approach - is it worth merging and supporting?

**Comment by time-less-ness** [comment]
I've been using this on a Mac M1 Max since the PR was raised and it's working fine for me. I've been hoping it will get merged so I can go back to compiling from `master` again. Really enjoying having 8k context.

**Comment by SlyEcho** [comment]
Let's merge and maybe then improve later.

**Comment by FNsi** [comment]
[another method?](https://github.com/cmp-nct/ggllm.cpp/issues/62#issuecomment-1636618089)

**Review by ggerganov — APPROVED** [review]

File: ggml.c [hunk 1]
```diff
@@ -15759,7 +15759,7 @@ static void ggml_compute_backward(struct ggml_context * ctx, struct ggml_tensor
             {
                 if (src0->grad) {
                     assert(src1->type == GGML_TYPE_I32);
-                    assert(ggml_nelements(src1) == 4);
+                    assert(ggml_nelements(src1) == 3);
```
Comment (Cebtenzzre) [inline ggml.c:15762]: Shouldn't this be 6? Based on the code immediately after it should be at least 4, I think, not 3.

**Comment by Green-Sky** [comment]
> Is there any reason why the following lines are unmodified and **still** use the hardcoded 10000.0 and 1.0 rope frequency and scale?


for the `ggml.c` lines, those appear to be the rope **backwards** passes, confusingly named `forward_rope_back`

**Comment by jxy** [comment]
I left the backward code untouched because I wasn't sure how I could correctly modify it and test it.

I'm also not sure about cuda bits.

**Comment by SlyEcho** [comment]
The CUDA part is broken right now, it should be fixed.

File: ggml.c (hunk 1 above)
Comment (DifferentialityDevelopment) [reply ggml.c:15762]: I went through the code an I also can't see why it's 3 when the lines just below it show it clearly taking 4 elements and looks like it designed to fail the assertion

Comment (Cebtenzzre) [reply ggml.c:15762]: @ggerganov

Comment (ggerganov) [reply ggml.c:15762]: Should be fixed in https://github.com/ggerganov/llama.cpp/commit/513f8619535a64fa9ace808cdcbcf66211535f5c

**Comment by bilal-aamer** [comment]
How do I implement this with RoPE and without it with current LLMs?

**Comment by abc-nix** [comment]
> How do I implement this with RoPE and without it with current LLMs?

You can read a bit more about RoPE use in llama.cpp in the [llama.cpp/examples/main/README.md](https://github.com/ggerganov/llama.cpp/tree/master/examples/main#extended-context-size)

Though I would recommend you try out the new Self-Extend support added in commit #4815 which I think is better, as you don't need to retrain the model to get better results.

**Comment by bilal-aamer** [comment]
> > How do I implement this with RoPE and without it with current LLMs?
> 
> You can read a bit more about RoPE use in llama.cpp in the [llama.cpp/examples/main/README.md](https://github.com/ggerganov/llama.cpp/tree/master/examples/main#extended-context-size)
> 
> Though I would recommend you try out the new Self-Extend support added in commit #4815 which I think is better, as you don't need to retrain the model to get better results.

Thanks @abc-nix!

What about the [implementation of customized RoPE](https://github.com/ggerganov/llama.cpp/discussions/1965#discussioncomment-6395555)

**Comment by abc-nix** [comment]
Sorry, @bilal-aamer, I am not sure what you are trying to ask here.

This PR adds customized RoPE support. Latter, [YaRN](https://github.com/jquesnelle/yarn) RoPE scaling was added in PR #2268 and some other fixes were added after that.

main's help has this to say about how the options and parameter to make use of RoPE/YaRN:

```
  --rope-scaling {none,linear,yarn}
                        RoPE frequency scaling method, defaults to linear unless specified by the model
  --rope-scale N        RoPE context scaling factor, expands context by a factor of N
  --rope-freq-base N    RoPE base frequency, used by NTK-aware scaling (default: loaded from model)
  --rope-freq-scale N   RoPE frequency scaling factor, expands context by a factor of 1/N
  --yarn-orig-ctx N     YaRN: original context size of model (default: 0 = model training context size)
  --yarn-ext-factor N   YaRN: extrapolation mix factor (default: 1.0, 0.0 = full interpolation)
  --yarn-attn-factor N  YaRN: scale sqrt(t) or attention magnitude (default: 1.0)
  --yarn-beta-slow N    YaRN: high correction dim or alpha (default: 1.0)
  --yarn-beta-fast N    YaRN: low correction dim or beta (default: 32.0)
```

I am not sure what you are trying to achieve or what exactly you are asking. Hopefully someone else isn't as obtuse as me and can help you out.

**Comment by madiarabis** [comment]
Is there any documentation on how to implement this or an example? I am kind of new in the field and I am fine tuning code llama 2 and I want to increase the context length. But between all these posts I am sort of confused how to implement it actually

**Pull request merged**